File: src/binarytree.py
class BinaryTree:
    def __init__(self, data=None):
        self.data = data
        self.left = None
        self.right = None

    @property
    def data(self):
        return self.__data

    @data.setter
    def data(self, data):
        self.__data = data

    def add_rightchild(self, node):
        if node is None:
            self.right = node
        elif self.data is None or node.data is None:
            self.right = node
        elif type(self.data) is type(node.data):
            self.right = node
        else:
            raise TypeError("Type mismatch between " + str(type(self.data))[8:-2] + " and " + str(type(node.data))[8:-2])

    def add_leftchild(self, node):
        if node is None:
            self.left = node
        elif self.data is None or node.data is None:
            self.left = node
        elif type(self.__data) is type(node.data):
            self.left = node
        else:
            raise TypeError("Type mismatch between " + str(type(self.data))[8:-2] + " and " + str(type(node.data))[8:-2])

    def preorder(self):
        if self.data is not None:
            yield self.data
        if self.left is not None:
            for d in self.left.preorder():
                yield d
        if self.right is not None:
            for d in self.right.preorder():
                yield d

    def __iter__(self):
        return self.preorder().__iter__()

File: src/BinaryTree.java
import java.util.Iterator;
import java.util.NoSuchElementException;
import java.util.Stack;

public class BinaryTree<T> implements Iterable<T> {

    T data;
    BinaryTree<T> left;
    BinaryTree<T> right;

    @Override
    public Iterator<T> iterator() {
        return new TreeIterator(this);
    }

    private class TreeIterator implements Iterator<T> {

        private Stack<BinaryTree<T>> stack = new Stack<>();

        TreeIterator(BinaryTree<T> tree) {
            if (tree.data != null)
                stack.push(tree); // starts with the root node
        }

        @Override
        public boolean hasNext() {
            return !stack.isEmpty();
        }

        @Override
        public T next() {
            if (!hasNext())
                throw new NoSuchElementException();
            BinaryTree<T> tree = stack.pop();
            if (tree.right != null)
                stack.push(tree.right);
            if (tree.left != null)
                stack.push(tree.left);
            return tree.data;
        }
    }

    public BinaryTree(T data)
    {
        this.data = data;
    }

    public BinaryTree() {}

    public void addLeftChild(BinaryTree<T> node)
    {
        left = node;
    }

    public void addRightChild(BinaryTree<T> node)
    {
        right = node;
    }
}
